third-party/bsp/stmf4cube/stm32f4_discovery/udc/stm32f4_udc.c

```diff
@@ -0,0 +1,372 @@
+
+		dev_addr = (uint8_t)(req->w_value) & 0x7FU;
+		HAL_PCD_SetAddress(&hpcd, dev_addr);
+		log_debug("addr=0x%x", dev_addr);
+
+		//TODO: create function(CtlSendStatus)
+		HAL_PCD_EP_Transmit(&hpcd, 0x00U, NULL, 0U);
+	}
+	else
+	{
+		HAL_PCD_EP_SetStall(&hpcd, 0x80U); //equivalent to CtlError()
+		HAL_PCD_EP_SetStall(&hpcd, 0U);
+	}
+}
+
+static void stm32f4_ll_set_configuration(struct usb_control_header *req) {
+	int config = req->w_value & 0xff;
+
+	log_debug("conf=0x%x", config);
+
+#if 1
+	HAL_PCD_EP_SetStall(&hpcd, 0x80U); //equivalent to CtlError()
+	HAL_PCD_EP_SetStall(&hpcd, 0U);
+#else
+	/*TODO: add check for config not found */
+	usb_gadget_set_config(stm32f4_udc.udc.composite, config);
+
+	HAL_PCD_EP_Transmit(&hpcd, 0x00U, NULL, 0U);
+#endif
+}
+
+static void stm32f4_ll_get_status(struct usb_control_header *req) {
+	uint16_t status = 0;
+
+	switch (req->bm_request_type & USB_REQ_RECIP_MASK) {
+	case USB_REQ_RECIP_DEVICE:
+		/*TODO: add check for w_length != 2bytes */
+
+		HAL_PCD_EP_Transmit(&hpcd, 0x00U, (uint8_t *) &status, 2U);
+
+		break;
+	/* TODO: add case for EPs recipient */
+	/* case USB_REQ_RECIP_ENDP: */
+	default:
+		log_error("Unsupported RECIP 0x%x",
+			req->bm_request_type & USB_REQ_RECIP_MASK);
+		HAL_PCD_EP_SetStall(&hpcd, req->bm_request_type & 0x80U);
+		break;
+	}
+}
+
+/**
+* @brief  stm32f4_ll_handle_standard_request
+*         Handle USB device requests
+* @param  req: USB request
+* @retval status
+*/
+static void stm32f4_ll_handle_standard_request(struct usb_control_header *req) {
+	int ret;
+
+	switch (req->b_request) {
+	case USB_REQ_SET_ADDRESS:
+		stm32f4_ll_set_address(req);
+		break;
+	case USB_REQ_SET_CONFIG:
+		stm32f4_ll_set_configuration(req);
+		break;
+	case USB_REQ_GET_CONFIG:
+		//stm32f4_ll_get_configuration(req);
+		log_debug("GET_CONFIGURATION");
+		break;
+	case USB_REQ_GET_STATUS:
+		stm32f4_ll_get_status(req);
+		break;
+	case USB_REQ_SET_FEATURE:
+		//stm32f4_ll_set_feature(req);
+		log_debug("SET_FEATURE");
+		break;
+	case USB_REQ_CLEAR_FEATURE:
+		//stm32f4_ll_clear_feature(req);
+		log_debug("CLEAR_FEATURE");
+		break;
+	default:
+		log_debug("usb:if\n");
+		ret = usb_gadget_setup(stm32f4_udc.udc.composite,
+			(const struct usb_control_header *) req, NULL);
+		if (ret != 0) {
+			log_debug("if_error\n");
+			log_error("Not implemented req 0x%x", req->b_request);
+			HAL_PCD_EP_SetStall(&hpcd, req->bm_request_type & 0x80U);
+		}
+		break;
+	}
+}
+
+/**
+ * @brief  Setup stage callback.
+ * @param  hpcd: PCD handle
+ * @retval None
+ */
+void HAL_PCD_SetupStageCallback(PCD_HandleTypeDef *hpcd) {
+	//TODO: nice to have a gadget state struct to contain all the info
+	struct usb_gadget_request *req;
+	int ret;
+
+	/* struct usb_control_header *ctrl = (struct usb_control_header *)hpcd->Setup; */
+	struct usb_control_header ctrl;
+	memcpy(&ctrl, hpcd->Setup, sizeof(struct usb_control_header));
+
+	log_debug("usb: setupstage\n");
```

Comment:
Why do no use `struct usb_control_header *ctrl = (struct usb_control_header *)hpcd->Setup` as you commented?